Add `write_record_metadata` to PyTorchFileWriter

mikaylagawarecki opened the pull request:
Add `PyTorchFileWriter.write_record_metadata(record_name, num_bytes)` that
- writes the zipfile header/end of central directory metadata for an entry*
- reserves `num_bytes` in the zipfile for the payload. 

*Since the payload is not provided, the CRC32 computation is skipped and 0s are written in the corresponding entry of the zipfile header

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125184



cc @albanD

mikaylagawarecki committed:
Add `write_record_metadata` to PyTorchFileWriter

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add `write_record_metadata` to PyTorchFileWriter"

Add `PyTorchFIleWriter.write_record_metadata(record_name, num_bytes)`, that will write the zipfile header/end of central directory metadata for an entry, and reserve `num_bytes` in the zipfile for the payload. 

Since the payload is not provided, the CRC32 computation is skipped and 0s are written in the corresponding entry of the zipfile header




[ghstack-poisoned]

albanD commented on caffe2/serialize/inline_container.cc:
```diff
@@ -612,6 +612,24 @@ size_t ostream_write_func(
   return ret;
 }
 
+// This seeking function pretends to behave like m_pWrite, but will actually
+// seek n bytes from the current offset
```
I'm not sure about this comment tbh. This is just doing the regular seek function?

From the write function above, the only thing I would have a comment about is the crc32 handling?

albanD commented on caffe2/serialize/inline_container.cc:
```diff
@@ -624,6 +642,15 @@ PyTorchStreamWriter::PyTorchStreamWriter(
   setup(archive_name_);
 }
 
+PyTorchStreamWriter::PyTorchStreamWriter(
```
Can't this have a default value for the seek_fn so that we can remove the overload above?

albanD commented on caffe2/serialize/inline_container.cc:
```diff
@@ -649,10 +676,22 @@ void PyTorchStreamWriter::setup(const string& file_name) {
     };
+    seek_func_ = [this](size_t nbytes) -> size_t {
+      file_stream_.seekp(nbytes, std::ios_base::cur);
+      return !file_stream_ ? -1L : static_cast<size_t>(file_stream_.tellp());
```
I'm a bit surprised by this conditions? If (!file_stream_), I guess the line above would have segfaulted? :p

albanD commented on caffe2/serialize/inline_container.cc:
```diff
@@ -649,10 +676,22 @@ void PyTorchStreamWriter::setup(const string& file_name) {
+
+  if (!seek_func_) {
+    seek_func_ = [=](size_t nbytes) -> size_t {
+      TORCH_CHECK(false, "seek func unimplemented");
```
I guess this is only used when using metadata-only? Should we make the error message specific for that use case?

albanD commented on test/test_serialization.py:
```diff
@@ -3969,6 +3969,49 @@ def test_serialization_dtype(self, dtype, weights_only):
+            # write nulls for 'data/0' and 'data/1'
+            with open(f if filename else f.name, 'rb+') as opened_f:
+                opened_f.seek(data_0_offset)
+                opened_f.write(b'\x00' * weight_nbytes)
```
Why not `b'0'` ?

albanD commented on third_party/miniz-2.1.0/miniz.c:
```diff
@@ -6308,7 +6309,10 @@ mz_bool mz_zip_writer_add_mem_ex_v2(mz_zip_archive *pZip, const char *pArchive_n
 	if (!(level_and_flags & MZ_ZIP_FLAG_COMPRESSED_DATA))
 	{
-		uncomp_crc32 = (mz_uint32)mz_crc32(MZ_CRC32_INIT, (const mz_uint8 *)pBuf, buf_size);
+		// Cannot compute crc_32 with no payload
```
I don't think "payload" is a concept here? I don't think you need a comment with the right variable names.

albanD commented on third_party/miniz-2.1.0/miniz.c:
```diff
@@ -6250,6 +6250,7 @@ mz_bool mz_zip_writer_add_mem_ex_v2(mz_zip_archive *pZip, const char *pArchive_n
     mz_uint32 extra_size = 0;
     mz_uint8 extra_data[MZ_ZIP64_MAX_CENTRAL_EXTRA_FIELD_SIZE];
     mz_uint16 bit_flags = 0;
+    mz_bool write_metadata = buf_size && !pBuf;
```
would say write_metadata_only

mikaylagawarecki committed:
Update on "Add `write_record_metadata` to PyTorchFileWriter"

Add `PyTorchFileWriter.write_record_metadata(record_name, num_bytes)` that
- writes the zipfile header/end of central directory metadata for an entry*
- reserves `num_bytes` in the zipfile for the payload. 

*Since the payload is not provided, the CRC32 computation is skipped and 0s are written in the corresponding entry of the zipfile header




[ghstack-poisoned]

albanD commented on caffe2/serialize/inline_container.h:
```diff
@@ -203,11 +203,21 @@ class TORCH_API PyTorchStreamReader final {
   explicit PyTorchStreamWriter(
-      const std::function<size_t(const void*, size_t)> writer_func);
+      const std::function<size_t(const void*, size_t)> writer_func,
+      const std::function<size_t(size_t)> seek_func = default_seek_func);
```
```suggestion
      const std::function<size_t(size_t)> seek_func = {});
```
If you don't want to deal with the default, std::function has an empty state that works very much like an optional and you can do `if (seek_func)` on it.

albanD reviewed (approved):
nit but SGTM!

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki committed:
Update on "Add `write_record_metadata` to PyTorchFileWriter"

Add `PyTorchFileWriter.write_record_metadata(record_name, num_bytes)` that
- writes the zipfile header/end of central directory metadata for an entry*
- reserves `num_bytes` in the zipfile for the payload. 

*Since the payload is not provided, the CRC32 computation is skipped and 0s are written in the corresponding entry of the zipfile header




cc albanD

[ghstack-poisoned]

mikaylagawarecki commented:
ghstack

mikaylagawarecki commented:
@pytorchbot merge

izaitsevfb commented:
[@pytorchbot](https://github.com/pytorchbot) revert -m "breaks internal builds, see D56962076" -c ghfirst

izaitsevfb commented:
@pytorchmergebot  revert -m "breaks internal builds, see D56962076" -c ghfirst

izaitsevfb commented:
@pytorchmergebot revert -m "breaks internal builds, see D56962076" -c ghfirst

albanD commented:
Closing this in favor of https://github.com/pytorch/pytorch/pull/125686 to reland myself.
